### PR TITLE
Move node_info_cache to common

### PR DESCRIPTION
**Pull request by bianpengyuan**
So that the code could be shared by all filters.

### CODE REVIEW

**Comment by bianpengyuan** [comment]
@mandarjog We'd better get this PR in as well, otherwise performance of sd filter would just suffer.

**Comment by kyessenov** [comment]
I'm not fully understanding the execution model of WASM nowadays. Is this cache shared between requests and is that safe to do so?

**Comment by bianpengyuan** [comment]
@kyessenov great point! This cache is shared between requests, but it could only be accessed by single thread and is non-blocking to access. So I think it should be safe. @jplevyak for confirmation, it should be safe to share state in root context between stream contexts, in both null and real wasm environment right?

**Comment by bianpengyuan** [comment]
@jplevyak can I get some advise on the question above? thanks!

**Comment by bianpengyuan** [comment]
@mandarjog @kyessenov Can I get your review on this? Need this to make sd filter performant..

File: extensions/stackdriver/stackdriver.cc [hunk 1]
```diff
@@ -152,31 +163,19 @@ FilterDataStatus StackdriverContext::onResponseBody(size_t body_buffer_length,
...
-  return dynamic_cast<StackdriverRootContext *>(root);
+StackdriverRootContext* StackdriverContext::getRootContext() {
+  RootContext* root = this->root();
+  return dynamic_cast<StackdriverRootContext*>(root);
 }
 
 void StackdriverContext::onLog() {
-  bool isOutbound = getRootContext()->isOutbound();
+  auto* root = getRootContext();
+  bool isOutbound = root->isOutbound();
```
Comment (mandarjog) [inline extensions/stackdriver/stackdriver.cc]: You can follow the stats filter here and move most of this login into the RootContext code.

**Comment by mandarjog** [comment]
@kyessenov  I don't think there any danger of cache being shared inadvertently here.
It is siloed per thread anyway. We really need to get this in. The header processing takes 30% CPU.  https://github.com/istio/istio/issues/18261
cc: @PiotrSikora

**Comment by bianpengyuan** [comment]
@kyessenov @mandarjog node info cache could be turned off now, let's get this in so that we could measure the planned performance optimization w or w/o cache?

File: extensions/common/node_info_cache.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+#endif  // NULL_PLUGIN
+
+namespace Wasm {
+namespace Common {
+
+const size_t DefaultNodeCacheMaxSize = 500;
+const wasm::common::NodeInfo EmptyNodeInfo;
+
+typedef std::shared_ptr<wasm::common::NodeInfo> NodeInfoPtr;
```
Comment (kyessenov) [inline extensions/common/node_info_cache.h]: nit: const NodeInfo

Comment (bianpengyuan) [reply extensions/common/node_info_cache.h]: Done

File: extensions/common/node_info_cache.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  }
+  auto nodeinfo_it = cache_.find(peer_id);
+  if (nodeinfo_it != cache_.end()) {
+    return nodeinfo_it->second;
+  }
+
+  // Do not let the cache grow beyond max_cache_size_.
+  if (int32_t(cache_.size()) > max_cache_size_) {
+    auto it = cache_.begin();
+    cache_.erase(cache_.begin(), std::next(it, max_cache_size_ / 4));
```
Comment (kyessenov) [inline extensions/common/node_info_cache.cc]: Just to confirm my understanding, the cache revocation strategy is to remove 1/4th of the elements at random?

Comment (bianpengyuan) [reply extensions/common/node_info_cache.cc]: Yes, there cache is not sophisticated enough. We should move to lru. @mandarjog wdyt?

Comment (mandarjog) [reply extensions/common/node_info_cache.cc]: Sure lru works. But this is a trivial implementation...

Comment (kyessenov) [reply extensions/common/node_info_cache.cc]: Yeah, let's not over-complicate until we can measure the effects first.
Can you add a test that exercises cache eviction? I don't see it in this PR.

Comment (bianpengyuan) [reply extensions/common/node_info_cache.cc]: Yeah, definitely not in this PR or 1.4 release, just something we could think about.

**Comment by bianpengyuan** [comment]
@kyessenov Any concerns on this change?

**Pull request merged**